### DISCUSSION TITLE
app.openapi() halts generation of api docs; endpoints declared after this don't get included in /docs or /redoc

### DISCUSSION THREAD

**Discussion by answerquest** [fastapi/fastapi#12279] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/path1")
async def path1():
    return {"message": "Hello World"}

def fetch_doc():
    apiD = app.openapi()

fetch_doc()

@app.get("/path2")
async def path2():
    return {"message": "Hello World 2"}
```

### Description

Hi, in my program I'm fetching the existing openapi doc at one point using `app.openapi()` function call. That code was being executed about midway through the program, before some of the api endpoints were declared.

I can understand that at this point the yet-to-be-declared endpoints would not be covered in the dict I get from `app.openapi()` - that is fine.

But I found that this also **halts** subsequent openapi generation entirely. The latter endpoints never get added to the doc if `app.openapi()` gets executed before they are declared.

Assuming the sample code given above is saved in "test1.py", command to run:
```
uvicorn test1:app --port 5900
```
Now upon visiting http://localhost:5900/docs in the browser, only "/path1" endpoint is visible. Same effect seen on http://localhost:5900/redoc as well.

![image](https://github.com/user-attachments/assets/b88c4a2d-c015-4d68-9c38-f69ba814b3aa)

But if we comment out the `fetch_doc()` function call, or move it to the bottom of the file after /path2 is declared and restart the program, then we now see both apis listed in the docs.

![image](https://github.com/user-attachments/assets/737e4f68-25dd-4dba-b379-de1dd9abbf16)

Not sure if this is a bug or a feature or an artefact of one of FastAPI's dependencies. I couldn't find its relation with Pydantic, and it's affecting both the /docs and /redoc so it's probably upstream from both Swagger-UI and ReDoc.

### Operating System

Linux

### Operating System Details

Linux Mint 21.1
Ubuntu equivalent: jammy / 22.04
cpu: x86_64

### FastAPI Version

0.115.0

### Pydantic Version

1.10.7

### Python Version

Python 3.11.2

### Additional Context

EDIT: looking up documentation entry: https://fastapi.tiangolo.com/reference/fastapi/?h=openapi%28%29#fastapi.FastAPI.openapi

> The first time it is called it stores the result in the attribute app.openapi_schema, and next times it is called, it just returns that same result. To avoid the cost of generating the schema every time.

Ok this may explain why the doc generation is halting. Still, I think it would be useful to have something out there informing developers that this halt happens if they call the function midway.

Secondary Question: Is there any way other than `app.openapi()` by which we can fetch the openapi json/dict generated by the Fastapi program from within the program, without having to, say, make an api call to self at the openapi json endpoint ? (I wanted to avoid doing that as it would require the code to know where it's hosted etc)

**Comment by sinisaos** [comment]
> Secondary Question: Is there any other way by which we can fetch the openapi json/dict generated by the Fastapi program from within program, without having to do, say, an api call to self at the openapi json endpoint ? (I wanted to avoid doing that as it would require the code to know where it's hosted etc)

@answerquest You can generate `openapi.json` from code without running the application [like in this answer](https://github.com/fastapi/fastapi/discussions/12216#discussioncomment-10689278). Sorry if I missed your point.

**Reply by answerquest** [reply]
@sinisaos thanks, but I've done the same thing already if you see my sample code. `app.openapi()` - you're using that only in your answer. 

The problem I've highlighted is that if you call it under say one api endpoint that comes before other api endpoints, then 
1) the output will not contain any of the subsequent apis whose declarations come after, and 
2) after you do this, the apidoc of the entire program is frozen at the moment at which you called `app.openapi()` . It will be missing the other apis permanently. 

This basically means that if I have say 100 api endpoints, then the only place where I can safely call  `app.openapi()` is the 100th endpoint's function, or in a separate function that is executed after all the endpoints have been declared/defined.

I think my use case is in programs that are ongoing-running whereas the answer you've given is about programs that won't run but you just want their openapi.

**Comment by sinisaos** [comment]
@answerquest I was just referring to the second part of your question and I think you need the openapi scheme without running the app. I was wrong and I'm sorry for it. As for the first part of the question, I thought you already answered that in the first part of your question. 

>The first time it is called it stores the result in the attribute app.openapi_schema, and next times it is called, it just returns that same result. To avoid the cost of generating the schema every time.

Cheers.

**Comment by YuriiMotov** [comment, marked as answer]
1) You already answered your own question about **halting**. `FastAPI.openapi` method generates openapi schema and caches it, then re-uses it on subsequent calls.

2) To generate always up-to-date schema you can use `get_openapi` (`from fastapi.openapi.utils import get_openapi`). It will not cache results and will not use cached results

**Comment by mwalsh161** [comment]
The endpoint that gets auto-generated for `app.openapi_url` dynamically adds `app.root_path` to `app.servers` prior to calling `app.openapi()`:

https://github.com/fastapi/fastapi/blob/1c7e2540c22ba5e81250cc715d82db31e52a63d6/fastapi/applications.py#L1083-L1089

The [documentation](https://fastapi.tiangolo.com/how-to/extending-openapi/) is really helpful, however it doesn't make it clear that this code-path produces different behavior than accessing it through the endpoint.

This creates a problem with calling `app.openapi` directly since this gets cached and the expected endpoint no longer produces the same result. Unfortunately, this is the behavior of other tools that interact with fastapi, like fastmcp:

https://github.com/jlowin/fastmcp/blob/18d4a6ab63cd8de50daaec600755eada10f1e551/src/fastmcp/server/server.py#L2553

Am I misunderstanding the intended usage or would it make sense to consolidate the `app.servers` and `app.root_path` logic so you get the same behavior through `app.openapi` and the endpoint?
